custom_components/octopus_intelligent/persistent_data.py

```diff
@@ -0,0 +1,89 @@
+"""Persistent data storage for the integration, based on the HASS helpers.storage.Store class."""
+import logging
+from dataclasses import asdict, dataclass, InitVar
+from typing import Any
+
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant
+from homeassistant.exceptions import IntegrationError
+from homeassistant.helpers.storage import Store
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class PersistentData:
+    """JSON-serialisable data persistence backed by the hass helpers.storage.Store class.
+
+    Frequently persisting data to "disk" can have undesired side effects when HASS is
+    running on some edge devices like a Raspberry Pi, whose storage consists of an SD card
+    that wears when data is written, leading to physical damage and data corruption.
+    For this reason, by default, the data is only saved when the HASS STOP event event
+    is fired, indicating that Home Assistant is quitting / restarting. This includes the
+    frontend web UI 'Restart' command, CTRL-C on the command line, the "docker container
+    stop" command, and generally when the SIGTERM signal is received by the HASS core
+    process.
+    """
+    # Note: InitVar fields are not persisted. They become arguments to __post_init__().
+    hass: InitVar[HomeAssistant]
+    account_id: InitVar[str]
+
+    # ---------------------------------------------
+    # Start of JSON-serialisable persistent fields.
+    #
+    last_seen_planned_dispatch_source: str = "smart-charge"
```

Comment:
This `smart-charge` default value only applies when the integration is first installed, or reinstalled after being deleted, when no previous data exists in persistent storage. Data read from persistent storage — even if it is an empty string — overrides this default value.

An alternative default value could be the empty string. If the default value was the empty string, and the Octopus API was responding with `"source": None` when the integration was installed or reinstalled, then issues #35 and #36 would apply until the Octopus API responded with a non-None `source` attribute. During this period, the integration sensors would not report that a smart charge was taking place, even if it was.

I think that this default value should be `smart-charge` for two reasons:

1. It means that issues [#35](https://github.com/megakid/ha_octopus_intelligent/issues/35) and [#36](https://github.com/megakid/ha_octopus_intelligent/issues/36) are _immediately_ resolved (subject to the [limitations](https://github.com/megakid/ha_octopus_intelligent/pull/37#issuecomment-1979168329) of the solution) when users upgrade the integration to a new version including the solution in this PR. Similarly, new users of this integration are more likely to find that it works straight away. (Otherwise, new users might simply decide that _"it doesn't work"_ and uninstall it.)

2. If one was to argue that this default value should be the empty string instead of `smart-charge`, then they’d also have to argue against the persistent storage of the `source` attribute more generally, because the persisted value — whether `smart-charge` or `bump-charge` — could be incorrect when the integration starts.

This goes back to the discussion of the corner-case involving bump charge in [another comment](https://github.com/megakid/ha_octopus_intelligent/pull/37#issuecomment-1979168329). It is possible, but unlikely, that a user will first install the integration while a bump charge is taking place ***and*** while the Octopus API is responding with `"source": None`. I think it is better to accept a temporary misreporting of a smart charge (when it is actually a bump charge) in this unlikely scenario than to misreport more often to more users (failing to report a smart charge while one is taking place).